File: pkg/strava/activities.go
package strava

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/gleich/lcp/pkg/secrets"
	"github.com/gleich/lumber/v2"
)

type Activity struct {
	Name      string    `json:"name"`
	SportType string    `json:"sport_type"`
	StartDate time.Time `json:"start_date"`
	Timezone  string    `json:"timezone"`
	Map       struct {
		SummaryPolyline string `json:"summary_polyline"`
	} `json:"map"`
	Trainer            bool    `json:"trainer"`
	Commute            bool    `json:"commute"`
	Private            bool    `json:"private"`
	AverageSpeed       float32 `json:"average_speed"`
	MaxSpeed           float32 `json:"max_speed"`
	AverageTemp        int32   `json:"average_temp,omitempty"`
	AverageCadence     float32 `json:"average_cadence,omitempty"`
	AverageWatts       float32 `json:"average_watts,omitempty"`
	DeviceWatts        bool    `json:"device_watts,omitempty"`
	AverageHeartrate   float32 `json:"average_heartrate,omitempty"`
	TotalElevationGain float32 `json:"total_elevation_gain"`
	MovingTime         uint32  `json:"moving_time"`
	SufferScore        float32 `json:"suffer_score,omitempty"`
	PrCount            uint32  `json:"pr_count"`
	Distance           float32 `json:"distance"`
	ID                 uint64  `json:"id"`
}

func FetchActivities(loadedSecrets secrets.Secrets) ([]Activity, error) {
	req, err := http.NewRequest("GET", "https://www.strava.com/api/v3/athlete/activities", nil)
	if err != nil {
		lumber.Error(err, "Failed to create new request")
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+loadedSecrets.StravaAccessToken)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		lumber.Error(err, "Failed to send request for Strava activities")
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		err = fmt.Errorf("received non-200 response: %d", resp.StatusCode)
		lumber.Error(err, "Failed to get a valid response from Strava")
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		lumber.Error(err, "Failed to read response body")
		return nil, err
	}

	var activities []Activity
	err = json.Unmarshal(body, &activities)
	if err != nil {
		lumber.Error(err, "Failed to parse JSON for Strava activities")
		lumber.Debug("Response body: ", body)
		return nil, err
	}

	return activities, nil
}

File: src/strava/activities.rs
use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use reqwest::Client;
use serde::{Deserialize, Serialize};

use crate::strava::token::TokenData;

#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
pub struct Activity {
    pub name: String,
    pub sport_type: String,
    pub start_date: DateTime<Utc>,
    pub timezone: String,
    pub map: Map,
    pub trainer: bool,
    pub commute: bool,
    pub private: bool,
    pub average_speed: f32,
    pub max_speed: f32,
    #[serde(default)]
    pub average_temp: i32,
    #[serde(default)]
    pub average_cadence: f32,
    #[serde(default)]
    pub average_watts: f32,
    #[serde(default)]
    pub device_watts: bool,
    #[serde(default)]
    pub average_heartrate: f32,
    pub total_elevation_gain: f32,
    pub moving_time: u32,
    #[serde(default)]
    pub suffer_score: f32,
    pub pr_count: u32,
    pub distance: f32,
    pub id: u64,
}

#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
pub struct Map {
    pub summary_polyline: String,
}

pub async fn fetch_recent(token_data: &TokenData, client: &Client) -> Result<Vec<Activity>> {
    let resp: reqwest::Response = client
        .get("https://www.strava.com/api/v3/athlete/activities")
        .bearer_auth(&token_data.access_token)
        .send()
        .await
        .context("sending request for recent activities failed")?;
    let resp_text = resp
        .text()
        .await
        .context("getting raw response text failed")?;
    let data: Vec<Activity> = serde_json::from_str(&resp_text).context(format!(
        "reading json failed from request to get activities: response: {}",
        resp_text
    ))?;
    Ok(data[0..6].to_vec())
}
